server/gitrest/src/utils/isomorphicgitManager.ts

```diff
@@ -0,0 +1,394 @@
+                        }),
+                        isomorphicGit.expandRef({
+                            fs: this.fileSystemManager,
+                            gitdir: this.directory,
+                            ref: refId,
+                        }),
+                    ]);
+                    return {
+                        resolvedRef,
+                        expandedRef,
+                    };
+                }));
+
+        return resolvedAndExpandedRefs.map(
+            (resolvedAndExpandedRef) =>
+                conversions.refToIRef(
+                    resolvedAndExpandedRef.resolvedRef,
+                    resolvedAndExpandedRef.expandedRef));
+    }
+
+    public async getRef(refId: string, externalWriterConfig?: IExternalWriterConfig): Promise<resources.IRef> {
+        try {
+            const [resolvedRef, expandedRef] = await Promise.all([
+                isomorphicGit.resolveRef({
+                    fs: this.fileSystemManager,
+                    gitdir: this.directory,
+                    ref: refId,
+                }),
+                isomorphicGit.expandRef({
+                    fs: this.fileSystemManager,
+                    gitdir: this.directory,
+                    ref: refId,
+                }),
+            ]);
+            return conversions.refToIRef(resolvedRef, expandedRef);
+        } catch (err) {
+            winston.error(`getRef error: ${safeStringify(err, undefined, 2)} repo: ${this.repoName} ref: ${refId}`);
+            return Promise.reject(err);
+        }
+    }
+
+    public async createRef(
+        createRefParams: resources.ICreateRefParams,
+        externalWriterConfig?: IExternalWriterConfig,
+    ): Promise<resources.IRef> {
+        await isomorphicGit.writeRef({
+            fs: this.fileSystemManager,
+            gitdir: this.directory,
+            ref: createRefParams.ref,
+            value: createRefParams.sha,
+        });
+        return conversions.refToIRef(createRefParams.sha, createRefParams.ref);
+    }
+
+    public async patchRef(
+        refId: string,
+        patchRefParams: resources.IPatchRefParams,
+        externalWriterConfig?: IExternalWriterConfig,
+    ): Promise<resources.IRef> {
+        await isomorphicGit.writeRef({
+            fs: this.fileSystemManager,
+            gitdir: this.directory,
+            ref: refId,
+            value: patchRefParams.sha,
+            force: true, // Isomorphic-Git requires force to be always true if we want to overwrite a ref.
+        });
+        return conversions.refToIRef(patchRefParams.sha, refId);
+    }
+
+    public async deleteRef(refId: string): Promise<void> {
+        try {
+            await isomorphicGit.deleteRef({
+                fs: this.fileSystemManager,
+                gitdir: this.directory,
+                ref: refId,
+            });
+        } catch (e: any) {
+            throw new NetworkError(500, `Failed to delete ref. Error: ${e}`);
```

Comment:
Out of curiosity: this is the only function where there is an explicit try..catch with a throw NetworkError. What is different about this one?

Reply:
I was trying to follow the same pattern as in the `NodeGitManager`:
https://github.com/microsoft/FluidFramework/blob/7a2e684e7bdf6e3111671d9b1ba38d64ec49e327/server/gitrest/src/utils/nodegitManager.ts#L295-L300

Deleting a ref in `nodegit` would return you a specific code depending on the result, and then `NodeGitManager` would throw the `NetworkError` based on that code.

[On the other hand, `isomorphic-git`'s `deleteRef` returns `Promise<void>`](https://isomorphic-git.org/docs/en/deleteRef). To keep behavior consistent with what `NodeGitManager` does, I used the try-catch to detect when a problem happens.

Reply:
Thanks for explaining!